Add client user guide for smithy java (#2522)
smithy-lang/smithy: docs/source-2.0/java/client/generating-clients.rst

        // Protocol implementations and auth schemes used by client
        implementation("com.example:my-protocol:1.0.0")
        implementation("com.example:my-auth-scheme:1.0.0")
    }

Now, define a service model in a ``model/`` directory at the root of your project.
The ``smithy-base``` Gradle plugin will automatically discover any models added to that directory.

---------------------------
Configuring code generation
---------------------------

In order to execute code generation, the ``java-client-codegen`` plugin must be added to
your :ref:`smithy-build <smithy-build>` config:

.. code-block:: diff
   :caption: smithy-build.json

    {
      "version": "1.0",
      "plugins": {
    +    "java-client-codegen": {
    +      "service": "com.example#CoffeeShop", // <- Replace with your service's ID
    +      // Generated Java code will use this as the root package namespace
    +      "namespace": "com.example.cafe"
    +    }
      }
    }

----------------------------------------
Add generated code to the Java sourceSet
----------------------------------------

Now the package is configured to generate client source code.

> However, the generated client code must be added to a `sourceSet <https://docs.gradle.org/current/dsl/org.gradle.api.tasks.SourceSet.html>`_

Reviewer: Should this have another preceding line break?

Author: I dont think so. Just needs wrapping to be fixed.

> to be compiled by Gradle. To add the generated client code to the ``main`` sourceSet update your Gradle build script to

Reviewer: ```suggestion
to be compiled by Gradle. To add the generated client code to the ``main`` sourceSet, update your Gradle build script to
```

==================
Generating Clients
==================

The Smithy Java :ref:`build plugin <plugins>`, ``java-client-codegen``, generates Java clients from Smithy models,
and can be executed with both `Gradle <https://gradle.org/>`_ (recommended) or the :ref:`Smithy CLI <smithy-cli>`.

.. admonition:: Important
    :class: note

>     The Smithy CLI is a prerequisite for this guide when using the ``smithy init``` commands.

Reviewer 2: ```suggestion
    The Smithy CLI is a prerequisite for this guide.
```

    See the :doc:`Smithy CLI installation guide <../../guides/smithy-cli/cli_installation>`
    if you do not already have the CLI installed.

-----------------------------------
Initial setup: Gradle (recommended)
-----------------------------------

> To generated a Java client for a service, first create a new Smithy Gradle project.

Reviewer 2: ```suggestion
To generate a Java client for a service, start by creating a new Smithy Gradle project.
```

.. tip::

    Use the ``smithy init``` CLI command to create a new Smithy project.
    The command ``smithy init -t quickstart-gradle`` will create a new

>     basic Smithy Gradle project.

Reviewer 2: Convert this to a short code-block, and briefly show the project structure like you did the in quickstart

> Apply the `smithy-base`_ plugin to your Gradle build script to build your Smithy model
> and execute build plugins.

Reviewer 2: ```suggestion
Apply the `smithy-base`_ plugin to your Gradle build script to build your Smithy model
and execute build plugins:
```

Reviewer 2: Check all of your statements before a code-block for a colon instead of a period.

        implementation("software.amazon.smithy.java:client-core:__smithy_java_version__")

        // Protocol implementations and auth schemes used by client
        implementation("com.example:my-protocol:1.0.0")
        implementation("com.example:my-auth-scheme:1.0.0")
    }

Now, define a service model in a ``model/`` directory at the root of your project.
The ``smithy-base``` Gradle plugin will automatically discover any models added to that directory.

---------------------------
Configuring code generation
---------------------------

In order to execute code generation, the ``java-client-codegen`` plugin must be added to
your :ref:`smithy-build <smithy-build>` config:

.. code-block:: diff
   :caption: smithy-build.json

    {
      "version": "1.0",
      "plugins": {
    +    "java-client-codegen": {
    +      "service": "com.example#CoffeeShop", // <- Replace with your service's ID
    +      // Generated Java code will use this as the root package namespace
    +      "namespace": "com.example.cafe"
    +    }
      }
    }

----------------------------------------
Add generated code to the Java sourceSet
----------------------------------------

> Now the package is configured to generate client source code. However, the generated client code must be added to
> a `sourceSet <https://docs.gradle.org/current/dsl/org.gradle.api.tasks.SourceSet.html>`_ to be compiled by Gradle.
> To add the generated client code to the ``main`` sourceSet, update your Gradle build script to include the following:

Reviewer 2: ```suggestion
Your package is now configured to generate Java client source code. However, the generated code must be
added to a `sourceSet <https://docs.gradle.org/current/dsl/org.gradle.api.tasks.SourceSet.html>`_ to be
compiled by Gradle. To add the generated code to the ``main`` sourceSet, add the following to your
Gradle build script:
```

.. code-block:: kotlin
   :caption: build.gradle.kts

    // Add generated Java sources to the main sourceSet so they are compiled alongside
    // any other Java code in your package
    afterEvaluate {
        val clientPath = smithy.getPluginProjectionPath(smithy.sourceProjection.get(), "java-client-codegen")
        sourceSets {
            main {
                java {
                    srcDir(clientPath)
                }
            }
        }
    }

    // Ensure client files are generated before java compilation is executed.
    tasks.named("compileJava") {
        dependsOn("smithyBuild")
    }

---------------
Generating code
---------------

> To generate and compile your client code run a build from the root of your Gradle project:

Reviewer 2: ```suggestion
To generate and compile your client code, run a build from the root of your Gradle project:
```

.. code-block:: sh

    ./gradlew clean build

> Building the project will generate code into the
> ``build/smithy-projections/<project-name>/source/java-client-codegen/`` directory.

Reviewer 2: ```suggestion
Building the project will generate code in the
``build/smithy-projections/<project-name>/source/java-client-codegen/`` directory.
```
